lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+open Lambda
+
+type error =
+  |  Ambiguous_constructor_arguments of lambda list
+
+exception Error of Location.t * error
+
+(** TMC (Tail Modulo Cons) is a code transformation that
+    rewrites transformed functions in destination-passing-style, in
+    such a way that certain calls that were not in tail position in the
+    original program become tail-calls in the transformed program.
+
+    As a classic example, the following program
+    {|
+     let[@tail_mod_cons] rec map f = function
+     | [] -> []
+     | x :: xs ->
+       let y = f x in
+       y :: map f xs
+    |}
+    becomes (expressed in almost-source-form; the translation is in
+    fact at the Lambda-level)
+    {|
+     let rec map f = function
+     | [] -> []
+     | x :: xs _>
+       let y = f x in
+       let dst = y :: Placeholder in
+       map_dps dst 1 f xs; dst
+     and map_dps dst offset f = function
+     | [] ->
+       dst.offset <- []
+     | x :: xs ->
+       let y = f x in
+       let dst' = y :: Placeholder in
+       dst.offset <- dst';
+       map_dps dst 1 f fx
+    |}
+
+    In this example, the expression (y :: map f xs) had a call in
+    non-tail-position, and it gets rewritten into tail-calls. TMC
+    handles all such cases where the continuation of the call
+    (what needs to be done after the return) is a "construction", the
+    creation of a (possibly nested) data block.
+
+    The code transformation generates two versions of the
+    input function, the "direct" version with the same type and
+    behavior as the original one (here just [map]), and
+    the "destination-passing-style" version (here [map_dps]).
+
+    Any call to the original function from outside the let..rec
+    declaration gets transformed into a call into the direct version,
+    which will itself call the destination-passing-style versions on
+    recursive calls that may benefit from it (they are in tail-position
+    modulo constructors).
+
+    Because of this inherent code duplication, the transformation may
+    not always improve performance. In this implementation, TMC is
+    opt-in, we only transform functions that the user has annotated
+    with an attribute to request the transformation.
+*)
+
+type 'offset destination = {
+  var: Ident.t;
+  offset: 'offset;
+  loc : Debuginfo.Scoped_location.t;
+}
+and offset = lambda
```

Comment:
Defining the offset type as an abstract type in a module might be worth it for later maybe turning it into `Int of int | Lambda of lambda` or something similar

Reply:
Something like:
```OCaml
module Offset :
sig
  type offset
  val add_dst_args : offset destination -> lambda list -> lambda list
  val assign_to_dst : offset destination -> lambda -> lambda
  val of_int : int -> offset
  val of_var : Ident.t -> offset
end = struct

type offset = lambda
(** In the OCaml value model, interior pointers are not allowed.  To
    represent the "placeholder to mutate" in DPS code, we thus use a pair
    of the block containing the placeholder, and the offset of the
    placeholder within the block.

    In the common case, this offset is an arbitrary lambda expression, typically
    a constant integer or a variable. We define ['a destination] as parametrized
    over the offset type to represent formal destination parameters (where
    the offset is an Ident.t), and maybe in the future statically-known
    offsets (where the offset is an integer).
*)

let add_dst_args ({var; offset} : offset destination) args =
  Lvar var :: offset :: args

let assign_to_dst {var; offset; loc} lam =
  Lprim(Psetfield_computed(Pointer, Heap_initialization),
        [Lvar var; offset; lam], loc)

let of_int pos = Lconst (Const_base (Const_int pos))
let of_var var = Lvar var
end
open Offset
```